docs/proposals/20230407-managed-k8s-capi-contract-changes.md

```diff
@@ -0,0 +1,380 @@
+
+  // ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+  // +optional
+  ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+  ...
+}
+
+
+type GCPManagedClusterSpec struct {
+  // Project is the name of the project to deploy the cluster to.
+  Project string `json:"project"`
+
+  // The GCP Region the cluster lives in.
+  Region string `json:"region"`
+
+  // NetworkSpec encapsulates all things related to the GCP network.
+  // +optional
+  Network NetworkSpec `json:"network"`
+
+  // FailureDomains is an optional field which is used to assign selected availability zones to a cluster
+  // FailureDomains if empty, defaults to all the zones in the selected region and if specified would override
+  // the default zones.
+  // +optional
+  FailureDomains []string `json:"failureDomains,omitempty"`
+
+  ...
+}
+```
+
+**Pros**
+
+- Simplifies provider implementation when reporting `ControlPlaneEndpoint`
+- Clearer separation between the lifecycle management of the general cloud infrastructure required for the cluster and the actual managed control plane (GKE in this example)
+- Follows the original intentions of an "infrastructure" and "control-plane" provider
+- Enables removal/addition of properties for a Managed Kubernetes cluster that may be different from a self-managed Kubernetes cluster
+- Works with ClusterClass
+
+**Cons**
+
+- Requires changes upstream to CAPI controllers to support the change of reporting `ControlPlaneEndpoint`
+- Duplication of API definitions between self-managed and managed `<Infra>Cluster` definitions and related controllers
+- Users need to be aware of when to use the unmanaged or managed `<Infra>Cluster` definitions.
+
+#### Flow 2: Change CAPI to make `<Infra>Cluster` optional
+
+This option follows along from the first flow above (`ControlPlaneEndpoint` reported by `<Infra>ControlPlane` resource rather than `<Infra>Cluster` resource), but takes it further and makes the `<Infra>Cluster` resource optional.
+
+This option would allow providers to implement only a `<Infra>ControlPlane` resource. Using CAPG as an example, rather than:
+
+- `Cluster` ←→ `GCPManagedCluster` + `GCPManagedControlPlane`
+
+We would enable:
+
+- `Cluster` ←→ `GCPManagedControlPlane`
+
+This would have the advantage of imposing a separation of configuration between each provider’s `<Infra>Cluster` and `<Infra>ControlPlane`’s resources. Because our observations have been that various Managed Kubernetes service providers do things a little bit differently, this separation is hard to define and enforce across all providers in a way that is agreeable to each provider.
+
+In practice this will help those providers who have implemented a `<Infra>Cluster` resource (e.g., `AzureManagedCluster` and `AWSManagedCluster`) as a sort of proxy resource that exists solely to fulfill the CAPI requirement for an `<Infra>Cluster` partner of its corresponding Cluster resource:
```

Comment:
Yes this would help CAPA by allowing the removal of `AWSManagedCluster`. 

But CAPA would still remain in a less than ideal place with its EKS implementation, as it does have extra infra. CAPA will find a way to evolve to the recommendations for the managed k8s proposal, so that it becomes consistent with CAPG, CAPOCI and CAPI clusters in general (i.e. separation of infra and cp).

So we need to stress that this option is only for implementations where there is no extra infra. It just so happens that for historic reasons that CAPA will fall into this.

Reply:
Makes sense, modified.